### ISSUE TITLE
Performance issue on limit sort combination #13379

### ISSUE THREAD

**Issue by tim-hanssen** [neo4j/neo4j#13380]
We are experiencing a very strange with the performance on queries with low limits. If the limit goes above 1024 the query is very fast, below its very slow:

- Neo4j version: 4.4.29
- Operating system: Ubuntu 18/22
- API/Driver: Neo4j Browser (but the same thing using any BOLT driver.)

We are running this on a core cluster members browser (not the leader), using BOLT, without any routing.

```
PROFILE MATCH (environment:Environment{Id: "3s736b3-08b3-4c0c-ae2e-33304faf28" })-[:CREATED_ASSET]->(asset:Asset)
RETURN asset ORDER BY asset.ChangedOn DESC LIMIT 10
```

`Started streaming 10 records after 7 ms and completed after 1915 ms.`

![plan (1)](https://github.com/neo4j/neo4j/assets/11192051/4b6b3bba-93ca-45b1-aa1e-8b93ecf7502e)

If we run the exact same query with a 10K limit, it's way faster:

```
PROFILE MATCH (environment:Environment{Id: "3s736b3-08b3-4c0c-ae2e-33304faf28" })-[:CREATED_ASSET]->(asset:Asset)
RETURN asset ORDER BY asset.ChangedOn DESC 10000
```

`Started streaming 10000 records after 31 ms and completed after 45 ms.`

![plan](https://github.com/neo4j/neo4j/assets/11192051/8953fddc-cbfa-4424-8c39-be17a6bc67d2)

The query plans follow the same route. In our testing we find out that the limit is around 1024. After that the query speeds drops to the fast levels.

**Comment by pontusmelke** [comment]
@tim-hanssen Thanks for reporting this. Do you think you could share the text version of the plans instead. In browser you click `Export TXT` in the upper-right corner and then you can just paste it here.

Regards
Pontus

**Comment by tim-hanssen** [comment]
Hey @pontusmelke I think that export function is only added in 5.x? We are running 4.4.29, and I don't see that export function.

**Comment by rweverwijk** [comment]
@tim-hanssen you can use https://browser.neo4j.io/ to connect to your local instance. Then you will have the latest browser version.

**Comment by tim-hanssen** [comment]
@rweverwijk 

![image](https://github.com/neo4j/neo4j/assets/11192051/7f82b7d0-71b2-4ede-bd76-8d3e6fb0ec76)

No difference to my local browser.

**Comment by pontusmelke** [comment]
Ok, sorry. I forgot when that was added. Browser will only display it if the server can render the plan description and I guess that wasn't added in 4.4

**Comment by rweverwijk** [comment]
@tim-hanssen are you on the "plan" tab?

<img width="785" alt="image" src="https://github.com/neo4j/neo4j/assets/226342/bbb472be-e3fa-4af7-8030-18a308b5f099">

**Comment by pontusmelke** [comment]
Do you think you could connect to the server with cypher-shell?

**Comment by tim-hanssen** [comment]
Yes the plan tab haha, I think it looks at the server version. Il try cypher shell.

**Comment by tim-hanssen** [comment]
Limit on 10:

```
+---------------------------------------------------------------------------------------------------------+
| Plan      | Statement   | Version      | Planner | Runtime     | Time | DbHits  | Rows | Memory (Bytes) |
+---------------------------------------------------------------------------------------------------------+
| "PROFILE" | "READ_ONLY" | "CYPHER 4.4" | "COST"  | "PIPELINED" | 4815 | 2395022 | 10   | 7136           |
+---------------------------------------------------------------------------------------------------------+


+----------------------------+-------------------------------------------------------------+----------------+--------+---------+----------------+------------------------+-----------+----------------------+---------------------+
| Operator                   | Details                                                     | Estimated Rows | Rows   | DB Hits | Memory (Bytes) | Page Cache Hits/Misses | Time (ms) | Ordered by           | Other               |
+----------------------------+-------------------------------------------------------------+----------------+--------+---------+----------------+------------------------+-----------+----------------------+---------------------+
| +ProduceResults@prepr      | asset                                                       |             10 |     10 |     140 |              0 |                   49/0 |     0.834 | asset.ChangedOn DESC | In Pipeline 1       |
| |                          +-------------------------------------------------------------+----------------+--------+---------+----------------+------------------------+-----------+----------------------+---------------------+
| +Top@prepr                 | `asset.ChangedOn` DESC LIMIT 10                             |             10 |     10 |       0 |           6944 |                    0/0 |     4.710 | asset.ChangedOn DESC | In Pipeline 1       |
| |                          +-------------------------------------------------------------+----------------+--------+---------+----------------+------------------------+-----------+----------------------+---------------------+
| +Projection@prepr          | asset.ChangedOn AS `asset.ChangedOn`                        |            613 | 598712 | 1197424 |                |                        |           |                      | Fused in Pipeline 0 |
| |                          +-------------------------------------------------------------+----------------+--------+---------+----------------+------------------------+-----------+----------------------+---------------------+
| +Filter@prepr              | asset:Asset                                                 |            613 | 598712 |  598712 |                |                        |           |                      | Fused in Pipeline 0 |
| |                          +-------------------------------------------------------------+----------------+--------+---------+----------------+------------------------+-----------+----------------------+---------------------+
| +Expand(All)@prepr         | (environment)-[anon_0:CREATED_ASSET]->(asset)               |            613 | 598712 |  598744 |                |                        |           |                      | Fused in Pipeline 0 |
| |                          +-------------------------------------------------------------+----------------+--------+---------+----------------+------------------------+-----------+----------------------+---------------------+
| +NodeUniqueIndexSeek@prepr | UNIQUE environment:Environment(Id) WHERE Id = $autostring_0 |              1 |      1 |       2 |            112 |              1583665/0 |  4763.809 |                      | Fused in Pipeline 0 |
+----------------------------+-------------------------------------------------------------+----------------+--------+---------+----------------+------------------------+-----------+----------------------+---------------------+
```

`ready to start consuming query after 5 ms, results consumed after another 4810 ms`

Limit on 1500

```
+---------------------------------------------------------------------------------------------------------+
| Plan      | Statement   | Version      | Planner | Runtime     | Time | DbHits  | Rows | Memory (Bytes) |
+---------------------------------------------------------------------------------------------------------+
| "PROFILE" | "READ_ONLY" | "CYPHER 4.4" | "COST"  | "PIPELINED" | 49   | 2415307 | 1500 | 121408         |
+---------------------------------------------------------------------------------------------------------+


+----------------------------+-------------------------------------------------------------+----------------+--------+---------+----------------+------------------------+-----------+----------------------+---------------------+
| Operator                   | Details                                                     | Estimated Rows | Rows   | DB Hits | Memory (Bytes) | Page Cache Hits/Misses | Time (ms) | Ordered by           | Other               |
+----------------------------+-------------------------------------------------------------+----------------+--------+---------+----------------+------------------------+-----------+----------------------+---------------------+
| +ProduceResults@prepr      | asset                                                       |            613 |   1500 |   20425 |              0 |                 7026/0 |   126.664 | asset.ChangedOn DESC | In Pipeline 1       |
| |                          +-------------------------------------------------------------+----------------+--------+---------+----------------+------------------------+-----------+----------------------+---------------------+
| +Top@prepr                 | `asset.ChangedOn` DESC LIMIT 1500                           |            613 |   1500 |       0 |         121216 |                    0/0 |     5.907 | asset.ChangedOn DESC | In Pipeline 1       |
| |                          +-------------------------------------------------------------+----------------+--------+---------+----------------+------------------------+-----------+----------------------+---------------------+
| +Projection@prepr          | asset.ChangedOn AS `asset.ChangedOn`                        |            613 | 598712 | 1197424 |                |                        |           |                      | Fused in Pipeline 0 |
| |                          +-------------------------------------------------------------+----------------+--------+---------+----------------+------------------------+-----------+----------------------+---------------------+
| +Filter@prepr              | asset:Asset                                                 |            613 | 598712 |  598712 |                |                        |           |                      | Fused in Pipeline 0 |
| |                          +-------------------------------------------------------------+----------------+--------+---------+----------------+------------------------+-----------+----------------------+---------------------+
| +Expand(All)@prepr         | (environment)-[anon_0:CREATED_ASSET]->(asset)               |            613 | 598712 |  598744 |                |                        |           |                      | Fused in Pipeline 0 |
| |                          +-------------------------------------------------------------+----------------+--------+---------+----------------+------------------------+-----------+----------------------+---------------------+
| +NodeUniqueIndexSeek@prepr | UNIQUE environment:Environment(Id) WHERE Id = $autostring_0 |              1 |      1 |       2 |            112 |              1583665/0 |  4733.602 |                      | Fused in Pipeline 0 |
+----------------------------+-------------------------------------------------------------+----------------+--------+---------+----------------+------------------------+-----------+----------------------+---------------------+
```

`ready to start consuming query after 5 ms, results consumed after another 44 ms`

The dataset of our test run is about 500k items.

**Comment by tim-hanssen** [comment]
My first feeling would be that the metric returned is just wrong. But thats hard to really measure.

**Comment by pontusmelke** [comment]
> My first feeling would be that the metric returned is just wrong. But thats hard to really measure.

Which metric do you mean? The `results consumed after another 44ms` part?

**Comment by tim-hanssen** [comment]
Yes, because if you look at the time column, it looks like the limit 10 is just a bit slower on the first part. But maybe im reading it wrong.

**Comment by pontusmelke** [comment]
No I think you are completely right. I have been able to reproduce this surprising behaviour, will get back to you when I have understood better what is going on.

**Comment by pontusmelke** [comment]
Hi again,
I have looked at it closer now and it is (just) an issue with the metric being returned. I have done some benchmarking and the query actually gets faster with a lower limit. However, when the used limit exceeds 1000 the value presented in browser and cypher-shell is wrong. If you instead look in query.log you will see the actual time we did spend on the query. 

We will still make sure to fix so that browser and cypher-shell will present the correct value.
